### ISSUE TITLE
TodoMVC is not accessible

### ISSUE THREAD

**Issue by dylanb** [davideast/ng2do#9]
After watching the video, I can tell that this application contains all of the accessibility issues that the 1.x Angular.js TodoMVC application has.

http://unobfuscated.blogspot.com/2015/02/angularjs-accessibility-woeful-todomvc.html

I would like to help you fix these as I have already done for the 1.x version

http://unobfuscated.blogspot.com/2015/02/angularjs-accessibility-knocking-off-my.html

This will give us better insight into what we need to do to allow Angular.js 2.0 to fulfill on its promise of "never having to give you the DOM" while at the same time being accessible.

As it currently stands, I anticipate that some of the issues fixed in the above-mentioned blog series will break this. In particular, the focus management issues will be hairy but I have some ideas for how zones and the router can be leveraged to solve some of these problems in a declarative way.

I will be able to submit some PRs once the installation issues mentioned in other issues have been fixed

**Comment by davideast** [comment]
Hi Dylan,

This is a very, very, simple example. Accessibility was not one of the goals due to time and the current state of the Angular2.

This app will not hit the TodoMVC site in it's current state. Someone will most likely beat me to it as well.

If you would like to discuss accessibility and Angular2 I would take that over to the http://github.com/angular/angular repository. 

**Edit:** I'm closing because this app will soon be merged into the angular/angular repo. We can add improvements there if you'd like.

**Comment by dylanb** [comment]
ok, I accept that, &lt;soapbox&gt;however you should not create (even simple) examples that are not accessible. It helps to perpetuate the inaccessibility of web applications. Angular.js is a hugely influential framework. You guys have a responsibility to show developers what it means to be a good web developer.&lt;/soapbox&gt;

**Comment by marcysutton** [comment]
@davideast I'd love to work on an accessible Angular2 demo, if you'd like some help.

**Comment by davideast** [comment]
@dylanb I understand your point. Know though that this example is not about what is rendered to the page. This example is to demonstrate the new concepts and syntax. There will be a better attempt at an accessible demo. I hope no one believes that this hastily made app adheres to best practices.

@marcysutton That sounds great. I do not any big plans to iterate on this demo. The most I will add is routing and a better build process (when available). This mostly because I'd rather not mess with TodoMVC and work off my own base. I'll give you a ping when we're getting closer to that point. I'm working on the docs for Angular2 and I'd love to have content about accessibility.

**Comment by dylanb** [comment]
@davideast I also offer help in creating accessible examples. I think that you should not publish any examples that are not accessible because whether you intend it or not, these examples get used and the coding practices replicate. This is the only example of Angular 2.0 that exists today (that I know of), so everyone who is interested is looking at this code. Now is the time to act and we are here to help.

NOTE: the reason I am approaching you is because you asked to be kicked (in the YouTube video), consider yourself kicked :-)

**Comment by davideast** [comment]
@dylanb I will definitely take you up on that offer. There are other examples in the angular/angular repo and we are working on others. I would love to see an example that demonstrates accessibility. If you would like to fork the angular/angular repo and create an example I will work to make sure the PR gets reviewed. Also, if you'd like to submit a PR for this repo I would love to review that as well, but as you mentioned the install needs improving. I'll give you an update when this is fixed.

**Comment by davideast** [comment]
I should mention that the plan is to add this sample into the angular/angular repo. That's why I closed this issue. However, if you believe you can get it updated soon I will be more than happy to reopen it.

**Comment by dylanb** [comment]
@davideast I will look at the angular/angular repo samples

**Comment by dylanb** [comment]
@davideast I looked in the angular/angular repo and only saw a very simple hello_world example, where can I find the more complex examples?

**Comment by davideast** [comment]
Hey @dylanb,

I thought there were a few more examples added but I was mistaken. None of the samples will be complex because Angular 2 is still in a very early alpha. This is one of the reasons why I stated this repo is an example of syntax only. Angular 2 isn't ready for production apps yet.

Do you have any ideas of examples that demonstrate good accessibility? I can help guide you if you let me know where you'd like to start.

**Comment by dylanb** [comment]
@davideast A good place to start would be with the TodoMVC application. I have made the 1.3.x version accessible, my fork of the repo is here

https://github.com/dylanb/todomvc

and a running version of the application is here

http://dylanb.github.io/todomvc/index.html#/

Not sure where the forms stuff is, but an accessible (but ugly) forms validation example is here 

http://dylanb.github.io/bower_components/a11yfy/examples/form-validation.html

An accessible ARIA menu is here

http://dylanb.github.io/bower_components/a11yfy/examples/menu.html

Accessible gesture-supporting calendar date picker is here

http://dylanb.github.io/datepicker/datepicker.html

All of these are good candidates because they expose different sorts of challenges that Angular.js 2.0 is trying to solve

**Comment by marcysutton** [comment]
@davideast when you get the build instructions for this app updated I'd love to play around with the accessibility stuff!

**Comment by dylanb** [comment]
@davideast I have a commit of an example Angular 2 ARIA menu that I'd like your comments on https://github.com/dylanb/angular/commit/4435e5f74d9bf80273ab76b7afd46bdbde452a75

**Comment by dylanb** [comment]
@davideast I would like to submit the PR for the above-mentioned example but would like some help with the submission guidelines for Angular 2.0 - I could not find any. Also, I had some questions specifically about the <content> element and its use in conjunction with the decorators that I would like come help with

**Comment by davideast** [comment]
@dylanb, cc @marcysutton 

That sounds great! As you can see I'm finishing up some polishes for this example right now. I plan on getting it merged into angular/angular within this week. This leaves you with a few options:
1. Send the PR against this repo and have me copy the changes over to the angular/angular repo.
2. Send the PR against the angular/angular repo later on when I have merged into master.

Let me know which you prefer.

**Comment by dylanb** [comment]
@davideast  I'll submit to Angular/angular - however are there any particular submission guidelines there I should adhere to? Test coverage, coding standards etc.?

**Comment by davideast** [comment]
@dylanb There are e2e tests that need to be written, but I haven't looked into that just yet. Once I know more I'll provide you with all the info I can.

I'll give you a ping when the todo app is in the angular/angular repo.

**Comment by marcysutton** [comment]
I imagine the contributing rules will stay mostly the same, although there may be additional coding guidelines for Angular 2. Here are the legacy guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#rules

**Comment by davideast** [comment]
@dylanb, @marcysutton 

I've added the todo sample to the main angular/angular repo. I need to write some e2e tests before we can take a crack at making it accessible, but it's almost there!
